### PR TITLE
Add support for the OES_vertex_array_object extension

### LINKED ISSUE #204: Missing support for WebGL OES_vertex_array_object
VertexArrayObjects are common GL functionality, exposed in WebGL1 via the OES_vertex_array_object extension. Would it be possible to add support for this?

https://developer.mozilla.org/en-US/docs/Web/API/OES_vertex_array_object

### PR DESCRIPTION
**Pull request by erikmchut**
Closes #204.

### Overview

This PR adds support for the OES_vertex_array_object extension. Vertex Array Objects encapsulate state from vertex buffers and vertex attributes.

### Extension information
https://www.khronos.org/registry/webgl/extensions/OES_vertex_array_object/
https://www.khronos.org/opengl/wiki/Vertex_Specification#Vertex_Array_Object

### Rationale

This is incredibly helpful for developers that target WebGL, WebGL2, and OpenGL3+ development in the same codebase, since the latter two require VertexArrayObjects and using them requires different buffer management.

### Implementation

To implement this, I had to split the state for WebGLVertexAttributes into 'Object' data and 'Global' data. Object is the vertex data that is stored in VAOs, whereas Global data is the data that is not stored in VAOs. With this split, I implemented the extension to according to the spec, changing the WebGLRenderingContext to use the different state objects and allowing the extension to swap the state when a VAO buffer is bound.

### Testing

All WebGL compliance tests continue pass with this change, including `extensions_oes-vertex-array-object` and `extensions_oes-vertex-array-object-bufferData` which comprehensively test the feature.

### CODE REVIEW

File: src/native/procs.h [hunk 1]
```diff
@@ -126,3 +126,7 @@
 	PFNGLGETSTRINGPROC glGetString;
 	PFNGLGETERRORPROC glGetError;
 	PFNGLDRAWBUFFERSEXTPROC glDrawBuffersEXT;
+        PFNGLGENVERTEXARRAYSOESPROC glGenVertexArraysOES;
```
Comment (dhritzkiv) [inline src/native/procs.h:129]: Nit pick: can you fix the indentation for these lines here to be two fewer spaces?

**Review by dhritzkiv — COMMENTED** [review]
Thanks for this PR! Looks like a great contribution. I'm going to have to do some more reading of the code + consider the changes surrounding WebGLVertexAttribute.

For reference, the inclusion of this functionality has increased the number of tests run + passed from 16137 to 16210.

File: src/native/procs.h (hunk 1 above)
Comment (erikmchut) [reply src/native/procs.h:129]: Looks like this file has actual tab characters instead of spaces (my editor added spaces). Should I update the whole file to spaces?

Comment (dhritzkiv) [reply src/native/procs.h:129]: Right - I didn't notice. Let's keep it all tabs, then.

**Comment by erikmchut** [comment]
Also, here is a table showing the types of information in global scope vs. vertex array object scope:
https://www.khronos.org/registry/OpenGL/specs/gl/glspec46.core.pdf#page=607&zoom=100,0,173

(newer opengl spec, but a good way to visualize).

TL;DR:

Global data:
- All the glVertexAttrib*v style methods set global data
- The bind state for an ARRAY_BUFFER buffer

Object data (in a vertex array object):
- The bind state for an ELEMENT_ARRAY_BUFFER 
- The value of the bound ARRAY_BUFFER at the time of a call to `vertexAttribPointer`
- All of the data set in vertexAttribPointer
- The glEnableVertexAttribArray setting

File: src/native/procs.h (hunk 1 above)
Comment (erikmchut) [reply src/native/procs.h]: Sounds good. Updated here and in procs.cc

**Comment by dhritzkiv** [comment]
Thanks for continuing to add more commits! I likely won't get to this until the weekend, so feel free to make more improvements where necessary

**Comment by erikmchut** [comment]
No worries – I'm done making changes. Let me know if you have any questions about the structure once you get the chance to review.

**Comment by dhritzkiv** [comment]
@robertleeplummerjr hey, could I get another pair of eyes on this? I figured your GPU projects might benefit from / be affected by these changes, and would love your thumbs up.

I've yet to sit down to look at this fully myself (probably later this weekend)

**Comment by erikmchut** [comment]
Hi @dhritzkiv, friendly ping regarding the above. Let me know if I can give any additional information to help with the review.

**Comment by dhritzkiv** [comment]
Released as v4.9.0!

Sorry for the delay - I had to review it a few times as there were some bigger changes made than previous extension additions (but they were necessary). I've also never use this extension before, so it was new to me.

Very nice code. Thank you for the contribution to the library.

**Comment by erikmchut** [comment]
Hi Daniel, that's excellent – thank you! I'll be using the new version right away. Happy to make more contributions in the future if I (or the team at 8th Wall) runs into any issues or missing features.

**Comment by dhritzkiv** [comment]
Please do!

**Comment by dhritzkiv** [comment]
Btw, I forgot to publish to npm last night. It is up now!
